src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.cs

```diff
@@ -30,19 +30,21 @@ public sealed partial class SafeSocketHandle : SafeHandleMinusOneIsInvalid
         private int _closeSocketThread;
         private int _closeSocketTick;
 #endif
+        private readonly bool _ownsHandle;
         private int _ownClose;
 
         public SafeSocketHandle(IntPtr preexistingHandle, bool ownsHandle)
             : base(ownsHandle)
         {
+            _ownsHandle = ownsHandle;
             SetHandleAndValid(preexistingHandle);
         }
 
-        private SafeSocketHandle() : base(true) { }
+        private SafeSocketHandle() : base(ownsHandle: true) => _ownsHandle = true;
 
         private bool TryOwnClose()
         {
-            return Interlocked.CompareExchange(ref _ownClose, 1, 0) == 0;
+            return _ownsHandle && Interlocked.CompareExchange(ref _ownClose, 1, 0) == 0;
```

Comment:
@tmds, is this the right fix?  Otherwise we end up spinning infinitely while waiting for ReleaseHandle to be called, which it never will.

(@wfurt, this may be the cause of the previous issue I mentioned that led me to use fstat to ensure the file descriptor was actually a socket.)

Reply:
The fix looks good to me.

Reply:
There may be some things in `Socket.Dispose` also that we don't want to do when the handle is not owning the socket.

Reply:
> There may be some things in Socket.Dispose also that we don't want to do when the handle is not owning the socket.

Anything in particular?  You think I should just make Socket.Dispose a nop entirely if !ownsHandle?

Reply:
Yes, I think we should make it nop.